app/lib/identity_tijuana/member_sync.rb

```diff
@@ -78,6 +147,88 @@ def self.import_user(user_id, sync_id)
+        raise "Id member has no linked TJ users"
+      end
+
+      # Some scenarios to think about before changing this:
+      #
+      # (1) A member email address may not match a TJ user email
+      # address. This could be because of:
+      #    * The cleaner fixed a typo
+      #    * Un-normalised data in both (example@gmail.com vs
+      #      Exam.ple@gmail.com
+      #    * The member's email was changed through Id CRM UI
+      #
+      # (1) A member's email address changed in Id, as a result no
+      # associated TJ users have the same email address.
+      #
+      # (2) A member subscribes and becomes active. They then typo
+      # their email when signing a petition from a new browser. The
+      # newer typo'ed user is not yet active (although it will become
+      # active tomorrow morning) but it is subscribed.
+      #
+      # (3) A member has two users associated with it, one with the
+      # correct email and one with a typo. The typo'ed user gets
+      # updated because they sign another petition with the same typo
+      # and a new postcode.
+      #
+      # (4) A member subscribes using a typo'ed email, becomes active
+      # since they are a new member but unsubscribed because their
+      # email bounced. Their Id member record has the correct email
+      # because the Cleanser fixed it. Thus the member and associated
+      # user have different email addresses. The member later signs
+      # another petition using the correct email address.
+
+      # The points above means that we should matching based on email
+      # first if present, otherwise fall back looking at subscribed
+      # users, then users that first created (assuming most members
+      # will likely not typo their email address).
+      #
+      # In particular we can't rely on most-recently-updated, since
+      # typo'ed users can still get updated because of (3) above. We
+      # also can't rely on active, since it's laggy but also typo'ed
+      # users can be active after they first sign up.
+
+      ext_ids = external_ids.map { |id| id.external_id.to_i }
+      users = User.where(id: ext_ids).order(created_at: :asc)
+      if users.count != external_ids.count
+        raise "Not enough matching users for external ids: #{ext_ids}"
+      end
+
+      matching_emails = users.select { |user|
+        # Need to normalise the email to match the normalisation Id
+        # would do, but as-is, i.e. without any typo correction
+        #
+        # XXX shamefully copied from Cleanser
+        user_email = user.email.downcase
+        user_email = user_email.gsub(/\s/, '')
+        user_email = user_email.gsub('%40', '@') # remove html entities
+
+        user_email == member.email
+      }
+      if matching_emails.count > 0
+        return matching_emails.first
+      end
+
+      subscribed_users = users.select { |user| user.is_member }
+      if subscribed_users.count > 0
+        subscribed_users.first
```

Comment:
What does the `first subscribed` actually mean?
Is it the earliest subscription? If so then would we need to look into `user_activity_events`.

Reply:
A user who is currently subscribed and whose created_at date is earlier than all other currently user members.

Reply:
(the heuristic here is "created earlier is more likely to be correct")

Reply:
@ivanb777 is there anything to be done here from your POV?

Reply:
not really, I must have missed `order(created_at: :asc)` above